### DISCUSSION TITLE
Does not save last tool head loaded

### DISCUSSION THREAD

**Discussion by Rneil-ca** [TypQxQ/KTC#44] [Configuring]
HI,

I think I am missing something in my config as I have been able to get everything working, but the KTC does not seem to remember which toolhead it last loaded.  With this config in the init_gocde set to this:
init_gcode:
  {% if myself.selected_tool|default("none")|lower == ktc.TOOL_NONE|lower %}
    KTC_TOOLCHANGER_DISENGAGE TOOLCHANGER={myself.name}
    KTC_SET_ACTIVE_TOOL TOOL={ktc.TOOL_NONE}
    KTC_SET_STATE TOOLCHANGER={myself.name} STATE=READY
    RESPOND MSG="Toolchanger {myself.name} initialized unlocked, without tool"
  {% else %}
    KTC_TOOLCHANGER_ENGAGE TOOLCHANGER={myself.name} IGNORE_ENGAGED=True
    RESPOND MSG="Toolchanger {myself.name} initialized with tool {myself.selected_tool}"
  {% endif %}

I get the following error:
Unknown tool already mounted.Can't automatically deselect unknown tool before selecting new tool.

I have been able to get the KTC to work with the following init_gcode:
init_gcode:
    KTC_TOOLCHANGER_DISENGAGE TOOLCHANGER={myself.name}
    KTC_SET_ACTIVE_TOOL TOOL={ktc.TOOL_NONE}
    KTC_SET_STATE TOOLCHANGER={myself.name} STATE=READY
    RESPOND MSG="Toolchanger {myself.name} initialized unlocked, without tool"

So I figure I am missing something somewhere but have look through all the documentation and example configs and can not figure out what I am missing.

Thank you for all your hard work on this project!

Ryan

**Comment by TypQxQ** [comment]
That is strange because you use same code as me from what I can see.
Are you getting the same error regardless of if a tool is mounted or unmounted and do you get that error at startup?
![image](https://github.com/user-attachments/assets/1c0717ef-ae7a-49fc-94ce-0a0af4440ecd)

**Comment by Rneil-ca** [comment]
Hi,

Switched the config back to original and now it is initializing without error (think it just had to get that unknown tool out of the variables file), but I still get an error when I select the same tool.  I know how to get around this with the advanced commands, but think I am still missing something as the tool is sitting in the Engaged status.

states from ktc_variables:
[State]
ktc_toolchanger_default = {'selected_tool': '0'}
ktc_toolchanger_default_toolchanger = {'selected_tool': '0'}

Initialization
![image](https://github.com/user-attachments/assets/e5b1eac7-e5a6-4b53-aa3c-8a454d795ee7)

Selecting the same tool as initialized:
![image](https://github.com/user-attachments/assets/4bc956c4-f16b-4097-9754-df0409ece57a)

**Comment by TypQxQ** [comment]
Strange. Please attach klippy.log and ktc.log so I can backtrack.

**Reply by Rneil-ca** [reply]
I created a work around using a klipper macro that runs at startup and uses the advanced commands.  Each one of my toolheads has its own sensor so it can tell what is actually mounted and then makes that the active one.  So I am good with this, I actually like it better. No need to troubleshoot further.  If you do want to look into this further let me know and I will revert my code back to original and send you those files.
